Az.3.0.0 migration guide (#10461)
Azure/azure-powershell: documentation/migration-guides/Az.3.0.0-migration-guide.md

  Get-AzBatchNodeAgentSku -BatchContext $Context
  ```
  #### After
  ```powershell
  $Context = Get-AzBatchAccountKey -AccountName "ContosoBatchAccount"
  Get-AzBatchSupportedImage -BatchContext $Context
  ```

### Compute
- `New-AzDiskConfig`
  `UploadSizeInBytes` parameter is used instead of `DiskSizeGB` for `New-AzDiskConfig` when CreateOption is Upload
  #### Before
  ```powershell
  $diskconfig = New-AzDiskConfig -Location 'Central US' -DiskSizeGB 1023 -SkuName Standard_LRS -OsType Windows -CreateOption Upload -DiskIOPSReadWrite 500 -DiskMBpsReadWrite 8
  ```
  #### After
  ```powershell
  $diskconfig = New-AzDiskConfig -Location 'Central US' -UploadSizeInBytes 1023 * 1024 * 1024 * 1024 -SkuName Standard_LRS   -OsType Windows -CreateOption Upload -DiskIOPSReadWrite 500 -DiskMBpsReadWrite 8
  ```
	
### HDInsight
- `Get-AzHDInsightJobOutput`
  - Updated the `Get-AzHDInsightJobOutput` cmdlet to support granular role-based access to the storage key.
  - Users with HDInsight Cluster Operator, Contributor, or Owner roles will not be affected.
  - Users with only the Reader role will need to specify `DefaultStorageAccountKey` parameter explicitly.

>   #### Users with only Reader role for cmdlet   Get-AzHDInsightJobOutput

Reviewer: Is this format correct?

# Migration Guide for Az 3.0.0

This document describes the changes between the 2.0.0 and 3.0.0 versions of Az 

<!-- TOC -->

- [Migration Guide for Az 3.0.0](#migration-guide-for-az-300)
  - [Batch](#batch)
    - [`Get-AzBatchNodeAgentSku`](#get-azbatchnodeagentsku)
    - [Incompatibility with previous versions of `Az.Resources`](#previous-version-incompatibility-with-az.resources-module)
  - [Compute](#compute)
    - [`New-AzDiskConfig`](#new-azdiskconfig)
  - [HDInsight](#hdinsight)
    - [`Get-AzHDInsightJobOutput`](#get-azhdinsightjoboutput)
    - [`Add-AzHDInsightConfigValues`](#add-azhdinsightconfigvalues)
    - [`Disable-AzHDInsightMonitoring`](#disable-azhdinsightmonitoring)
    - [`Enable-AzHDInsightMonitoring`](#enable-azhdinsightmonitoring)
    - [`Get-AzHDInsightMonitoring`](#get-azhdinsightmonitoring)

>     - [`Get-HDInsightProperty`](#get-hdinsightproperty)

Reviewer 2: ```suggestion
    - [`Get-AzHDInsightProperty`](#get-azhdinsightproperty)
```

#### Before
```powershell
Enable-AzHDInsightOMS Enable-AzHDInsightMonitoring -Name testcluster -WorkspaceId 1d364e89-bb71-4503-aa3d-a23535aea7bd -PrimaryKey <key for workspace 1d364e89-bb71-4503-aa3d-a23535aea7bd>
```
```powershell
Enable-AzHDInsightOperationsManagementSuite Enable-AzHDInsightMonitoring -Name testcluster -WorkspaceId 1d364e89-bb71-4503-aa3d-a23535aea7bd -PrimaryKey <key for workspace 1d364e89-bb71-4503-aa3d-a23535aea7bd>
```

#### After
```powershell
Enable-AzHDInsightMonitoring Enable-AzHDInsightMonitoring -Name testcluster -WorkspaceId 1d364e89-bb71-4503-aa3d-a23535aea7bd -PrimaryKey <key for workspace 1d364e89-bb71-4503-aa3d-a23535aea7bd>
```

### `Get-AzHDInsightMonitoring`
Added a new `Get-AzHDInsightMonitoring` cmdlet. Use this cmdlet to get the status of monitoring installation in an Azure HDInsight cluster (replaces `Get-AzHDInsightOperationsManagementSuite` and `Get-AzHDInsightOMS`).

#### Before
```powershell
Get-AzHDInsightOMS -Name testcluster
```
```powershell
Get-AzHDInsightOperationsManagementSuite -Name testcluster
```

#### After
```powershell
Get-AzHDInsightMonitoring -Name testcluster
```

> ### `Get-HDInsightProperty`

Reviewer 2: ```suggestion
### `Get-AzHDInsightProperty`
```

Cmdlet `Get-HDInsightProperty` removed alias to `Get-AzHDInsightProperties`.

#### Before
Using deprecated alias
```powershell
Get-AzHDInsightProperties -Location "East US 2"
```

#### After
```powershell

> Get-HDInsightProperty -Location "East US 2"

Reviewer 2: ```suggestion
Get-AzHDInsightProperty -Location "East US 2"
```

```

### `Grant-AzHDInsightRdpServicesAccess`
Removed the `Grant-AzHDInsightRdpServicesAccess` and `Revoke-AzHDInsightRdpServicesAccess` cmdlets. These are no longer necessary because clusters using Windows OS type are not supported. Please create a cluster using Linux OS type instead.

### `Remove-AzHDInsightCluster`
The output type of `Remove-AzHDInsightCluster` changed from `Microsoft.Azure.Management.HDInsight.Models.ClusterGetResponse` to `bool`.

#### Before
```powershell
$cluster = Remove-AzHDInsightCluster -ClusterName "your-hadoop-001"
```

#### After
```powershell
Remove-AzHDInsightCluster -ClusterName "your-hadoop-001" -PassThru
True
```

### `Revoke-AzHDInsightRdpServicesAccess`
The cmdlet is deprecated. There is no replacement for it.

> ### `AzHDInsightGatewayCredential`

Reviewer 2: ```suggestion
### `Set-AzHDInsightGatewayCredential`
```

Reviewer 2: We should add more detail here (can be later this week).  What PowerShell users will be interested in is how the properties have changed,  What properties of the old object are no longer available in the new type, and how can I get equivalents from the new type?

#### Before
```powershell
Remove-AzIotHubEventHubConsumerGroup -ResourceGroupName "myresourcegroup" -Name "myiothub" -EventHubConsumerGroupName myconsumergroup -EventHubEndpointName "/EventHubEndpointName"
```

#### After
```powershell
Remove-AzIotHubEventHubConsumerGroup -ResourceGroupName "myresourcegroup" -Name "myiothub" -EventHubConsumerGroupName myconsumergroup
```

### `Set-AzIotHub`
Parameter `OperationsMonitoringProperties` is deprecated without being replaced as IotHub is no longer using built-in endpoint("operationsMonitoringEvents").
	

## RecoveryServices

### `Edit-AzRecoveryServicesAsrRecoveryPlan`
`ASRRecoveryPlanGroup.ReplicationProtectedItems`, `ASRRecoveryPlanGroup.StartGroupActions` and `ASRRecoveryPlanGroup.EndGroupActions` is removed from output.

### `Get-AzRecoveryServicesAsrRecoveryPlan`
`ASRRecoveryPlanGroup.ReplicationProtectedItems`, `ASRRecoveryPlanGroup.StartGroupActions` and `ASRRecoveryPlanGroup.EndGroupActions` is removed from output.

### `New-AzRecoveryServicesAsrReplicationProtectedItem`

> Parameter IncludeDiskId is changed in future release to support directly write to managed disk support Azure Site Recovery.

Reviewer 2: ```suggestion
Parameter IncludeDiskId is changed to support directly writing to a managed disk in Azure Site Recovery.
```

## ServiceFabric

### `Add-ServiceFabricApplicationCertificate`
Removed `Add-ServiceFabricApplicationCertificate` as this scenario is covered by `Add-AzVmssSecret`.

#### Before
```powershell
Add-AzServiceFabricApplicationCertificate -ResourceGroupName "Group1" -Name "Contoso01SFCluster" -SecretIdentifier "http://keyVaultName.vault.contoso.net/secrets/secretName/secretVersion"
```

#### After
```powershell
$Vault = Get-AzKeyVault -VaultName "ContosoVault"
$CertConfig = New-AzVmssVaultCertificateConfig -CertificateUrl "http://keyVaultName.vault.contoso.net/secrets/secretName/secretVersion" -CertificateStore "Certificates"
$VMSS = New-AzVmssConfig
Add-AzVmssSecret -VirtualMachineScaleSet $VMSS -SourceVaultId $Vault.ResourceId -VaultCertificate $CertConfig
```

## Sql

### `Get-AzSqlDatabaseSecureConnectionPolicy`
Note that secure connection is deprecated and so command is removed. Please use the SQL database blade in the Azure portal to view the connection strings

### `Get-AzSqlDatabaseIndexRecommendations`

> `Get-AzSqlDatabaseIndexRecommendations` alias is removed.

Reviewer 2: ```suggestion
`Get-AzSqlDatabaseIndexRecommendations` alias is removed. Use `Get-AzSqlDatabaseIndexRecommendation` instead.
```

### `Get-AzSqlDatabaseRestorePoints`

> `Get-AzSqlDatabaseRestorePoints` alias is removed.

Reviewer 2: ```suggestion
`Get-AzSqlDatabaseRestorePoints` alias is removed. Use `Get-AzSqlDatabaseRestorePoint` instead.
```

### `Get-AzSqlDatabaseAuditing`
- The cmdlet `Get-AzSqlDatabaseAudit` is replacing this cmdlet.

> - The output type is changing from the existing type :'Microsoft.Azure.Commands.Sql.Auditing.Model.DatabaseBlobAuditingSettingsModel' to the new type :'Microsoft.Azure.Commands.Sql.Auditing.Model.DatabaseAuditingSettingsModel'

Reviewer 2: ```suggestion
- The output type is changing from the existing type :'Microsoft.Azure.Commands.Sql.Auditing.Model.DatabaseBlobAuditingSettingsModel' to the new type :'Microsoft.Azure.Commands.Sql.Auditing.Model.DatabaseAuditingSettingsModel', removing properties `AuditState` and `StorageAccountName`. and `StorageAccountSubscriptionId`.  Scripts can retrieve storage account information from the new `StorageAccountResourceId` property. 
```

ServerName                          : server01
DatabaseName                        : database01
AuditAction                         : {}
ResourceGroupName                   : resourcegroup01
AuditActionGroup                    : {SUCCESSFUL_DATABASE_AUTHENTICATION_GROUP, FAILED_DATABASE_AUTHENTICATION_GROUP,
                                       BATCH_COMPLETED_GROUP, ...}
PredicateExpression                 : statement <> 'select 1'
BlobStorageTargetState              : Enabled
StorageAccountResourceId            : /subscriptions/7fe3301d-31d3-4668-af5e-211a890ba6e3/resourceGroups/resourcegroup01/providers/Microsoft.Storage/storageAccounts/mystorage
StorageKeyType                      : Primary
RetentionInDays                     : 0
EventHubTargetState                 : Enabled
EventHubName                        : eventHubName
EventHubAuthorizationRuleResourceId : EventHubAuthorizationRuleResourceId
LogAnalyticsTargetState             : Enabled
WorkspaceResourceId                 : "/subscriptions/4b9e8510-67ab-4e9a-95a9-e2f1e570ea9c/resourceGroups/insights-integration/providers/Microsoft.OperationalInsights/workspaces/viruela2"
```

> ### `Set-AzSqlDatabaseAudit`

Reviewer 2: ```suggestion
### `Set-AzSqlDatabaseAuditing`
```

- The cmdlet `Set-AzSqlDatabaseAudit` is replacing this cmdlet.
- The output type is changing from the existing type :'Microsoft.Azure.Commands.Sql.Auditing.Model.DatabaseBlobAuditingSettingsModel' to the new type :'bool'

#### Before
```powershell
Set-AzSqlDatabaseAuditing -State Enabled -ResourceGroupName "ResourceGroup01" -ServerName "Server01" -StorageAccountName "Storage22" -DatabaseName "Database01"
```

#### After
```powershell
Set-AzSqlDatabaseAudit -ResourceGroupName "ResourceGroup01" -ServerName "Server01" -DatabaseName "Database01" -BlobStorageTargetState Enabled -StorageAccountResourceId "/subscriptions/7fe3301d-31d3-4668-af5e-211a890ba6e3/resourceGroups/resourcegroup01/providers/Microsoft.Storage/storageAccounts/mystorage"
```

> ### `Get-AzSqlServerAudit`

Reviewer 2: ```suggestion
### `Get-AzSqlServerAuditing`
```

- The cmdlet `Get-AzSqlServerAudit` is replacing this cmdlet.

> - The output type is changing from the existing type :'Microsoft.Azure.Commands.Sql.Auditing.Model.DatabaseBlobAuditingSettingsModel' to the new type :'Microsoft.Azure.Commands.Sql.Auditing.Model.ServerAuditingSettingsModel'

Reviewer 2: ```suggestion
- The output type is changing from the existing type :'Microsoft.Azure.Commands.Sql.Auditing.Model.DatabaseBlobAuditingSettingsModel' to the new type :'Microsoft.Azure.Commands.Sql.Auditing.Model.ServerAuditingSettingsModel'.  Properties `AuditState`, `StorageAccountName`, and `StorageAccountSubscriptionId` are removed.  Scripts that use `StorageAccountName` and `StorageAccountSubscriptionId` proeprties can retrieve this information from the new `StorageAccountResourceId` property.
```

```

#### After
```powershell
PS C:\> Get-AzSqlServerAudit -ResourceGroupName "resourcegroup01" -ServerName "server01"
ServerName                          : server01
ResourceGroupName                   : resourcegroup01
AuditActionGroup                    : {SUCCESSFUL_DATABASE_AUTHENTICATION_GROUP, FAILED_DATABASE_AUTHENTICATION_GROUP, BATCH_COMPLETED_GROUP, ...}
PredicateExpression                 : statement <> 'select 1'
BlobStorageTargetState              : Enabled
StorageAccountResourceId            : /subscriptions/7fe3301d-31d3-4668-af5e-211a890ba6e3/resourceGroups/resourcegroup01/providers/Microsoft.Storage/storageAccounts/mystorage
StorageKeyType                      : Primary
RetentionInDays                     : 0
EventHubTargetState                 : Enabled
EventHubName                        : eventHubName
EventHubAuthorizationRuleResourceId : EventHubAuthorizationRuleResourceId
LogAnalyticsTargetState             : Enabled
WorkspaceResourceId                 : "/subscriptions/4b9e8510-67ab-4e9a-95a9-e2f1e570ea9c/resourceGroups/insights-integration/providers/Microsoft.OperationalInsights/workspaces/viruela2"
```

> ### `Set-AzSqlServerAudit`

Reviewer 2: ```suggestion
### `Set-AzSqlServerAuditing`
```
